[GStreamer] plugin scanner not available in bwrap sandbox

philn opened the pull request:
#### 526ec6619aef7474f938f9acec7a8b755a658e80
<pre>
[GStreamer] plugin scanner not available in bwrap sandbox
<a href="https://bugs.webkit.org/show_bug.cgi?id=287368">https://bugs.webkit.org/show_bug.cgi?id=287368</a>

Reviewed by Michael Catanzaro.

On UsrMerge systems where /lib64 is a symlink to /usr/lib64 and where the GStreamer registry
attempts to load the plugin-scanner from /lib64/ the path needs to be allow-listed in the sandbox.

* Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
(WebKit::bindGStreamerData):

Canonical link: <a href="https://commits.webkit.org/296779@main">https://commits.webkit.org/296779@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b6d2cdc65901f168c3fd793e48a94f5fd1c1f8b5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109604 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29262 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19691 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/115621 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59832 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29940 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37850 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83274 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112552 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23850 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98708 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63733 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23229 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59420 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93225 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16891 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/118414 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36643 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27129 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/92282 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37016 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94968 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92100 "run-api-tests-without-change (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23457 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37073 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14815 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32458 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36537 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42008 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36197 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39540 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37907 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro reviewed (approved):
I'm going to click Approve, but I do wonder if this is really significantly easier than just adding some more symlinks to `sandboxArgs` in `bubblewrapSpawn` to fix things for more than just GStreamer?

We could test if /lib64 is a symlink on the host, and if so, then do `"--symlink", "/lib64", "/usr/lib64",` and that should probably fix this? Then we don't need to wonder what else besides GStreamer might be broken.

I checked what flatpak does. `flatpak_run_setup_usr_links()` creates symlinks for `abs_usrmerged_dirs` which is defined in flatpak-exports.c to be /bin, /lib, /lib32, /lib64, and /sbin. That's not so many locations; doesn't seem too hard?

mcatanzaro commented on Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
```diff
@@ -390,6 +390,27 @@ static const char* environmentVariableValue(const char* name, const char* defaul
+        auto fileType = FileSystem::fileType(directory);
+        if (fileType == FileSystem::FileType::SymbolicLink) {
+            auto symLink = makeString("/usr"_s, directory, '/');
+            args.appendVector(Vector<CString> { "--symlink", directory.utf8(), symLink.utf8() });
```
I think this is backwards, which would explain why /usr/bin/xdg-dbus-proxy is missing:

```suggestion
            auto linkTarget = makeString("/usr"_s, directory, '/');
            args.appendVector(Vector<CString> { "--symlink", linkTarget.utf8(), directory.utf8() });
```

philn replied:
I tried both ways, didn't notice an improvement either way.

mcatanzaro commented on Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
```diff
@@ -390,6 +390,27 @@ static const char* environmentVariableValue(const char* name, const char* defaul
+            args.appendVector(Vector<CString> { "--symlink", directory.utf8(), symLink.utf8() });
+        } else if (fileType == FileSystem::FileType::Directory) {
+            auto symLink = makeString("/usr"_s, directory);
+            args.appendVector(Vector<CString> { "--ro-bind-try", symLink.utf8(), symLink.utf8() });
```
This fails on non-usrmerged systems, e.g. if /bin is not a symlink then it won't be bound.

What I would do is bind normally here instead:

```suggestion
            args.appendVector(Vector<CString> { "--ro-bind", directory.utf8(), directory.utf8() });
```

That will take care of binding any paths not under /usr, like /bin or /lib. But we also need to bind the paths under /usr, like /usr/bin or /usr/lib. We could do that above, in the `if (fileType == FileSystem::FileType::SymbolicLink)` condition, but that would fail if the symlinks ever get removed (you never know, it could happen 50 years from now!), so probably best to just restore the use of --ro-bind-try that you removed for the locations under /usr?

philn replied:
The reason the current wip patch is like this is that `--ro-bind*` and `--symlink` don't seem to play well together. If I keep both I get this `bwrap: Can't make symlink at /bin: destination exists and is not a symlink`.

I think this needs the attention of someone actually familiar with the `bwrap` options.

mcatanzaro replied:
The options do exactly what you tell them to. It's either create a symlink or mount a directory. The error message indicates exactly what is wrong: this is a usrmerged system and /bin is supposed to be a symlink to /usr/bin, so BubblewrapLauncher.cpp shouldn't be trying to --ro-bind the host /bin. But we do need to --ro-bind it on non-usrmerged systems. We need to do _one of_ create symlink or bind host directory -- never both -- for each of these locations. That error message indicates your code is doing both. It's probably a simple mistake?

mcatanzaro replied:
Also: I notice the existing sandbox intentionally skips binding /bin except in the `enableDebugPermissions` case. We should probably maintain that behavior and just altogether remove /bin and /sbin from this function. We could create the symlinks in the `enableDebugPermissions` codepath if you want, though.

philn replied:
You realize you contradict yourself then? You said "so probably best to just restore the use of --ro-bind-try that you removed for the locations under /usr?" and now "We need to do one of create symlink or bind host directory -- never both -- for each of these locations".

mcatanzaro replied:
Use --ro-bind-try in the non-usrmerge condition, when /bin is not a symlink.

philn committed:
[GStreamer] plugin scanner not available in bwrap sandbox
https://bugs.webkit.org/show_bug.cgi?id=287368

Reviewed by Michael Catanzaro.

On UsrMerge systems where /lib64 is a symlink to /usr/lib64 and where the GStreamer registry
attempts to load the plugin-scanner from /lib64/ the path needs to be allow-listed in the sandbox.

* Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
(WebKit::bindGStreamerData):

Canonical link: https://commits.webkit.org/296779@main
